evproc/processor.py

```diff
@@ -308,11 +311,16 @@ def process(self, ev):
                   ``StopProcessing`` exception raised by a processor.
         """
 
-        # Walk through all the processors and process the event
-        for proc in self._get_procs(ev):
-            try:
-                proc(ev)
-            except exc.StopProcessing as ex:
-                return ex.retval
+        event_deque = collections.deque([ev])
+        while event_deque:
+            ev = event_deque.popleft()
+            # Walk through all the processors and process the event
+            for proc in self._get_procs(ev):
+                try:
+                    additional_events = proc(ev)
+                except exc.StopProcessing as ex:
+                    return ex.retval
```

Comment:
Good observation. I hadn't thought of this case. I think this behavior (i.e. event2 and event3 aren't processed) is reasonable.

An alternative would be to stop calling any additional event1 handlers (if such exist), but to continue processing the event2 and event3 returned by proc1. This seems like a perfectly plausible alternative, except that in this version I don't know what should be returned if multiple processors raise StopProcessing (e.g. one for event1, one for event2, and one for event3). If we really wanted to handle it this way, I suppose we could "yield ex.retval" rather than "return ex.retval" each time a handler raises StopProcessing, though this may be a bit odd (though has the advantage of occurring at the right time, e.g. we'd yield the ex.retval from proc2 before processing event2 and event3).